arangod/IResearch/IResearchView.cpp

```diff
@@ -1573,78 +1743,40 @@ arangodb::Result IResearchView::link(
   builder.close();
 
   std::unordered_set<TRI_voc_cid_t> collections;
-  auto result = updateLinks(collections, *vocbase, *this, builder.slice());
-
-  if (result.ok()) {
-    WriteMutex mutex(_mutex); // '_meta' can be asynchronously read
-    SCOPED_LOCK(mutex);
-
-    collections.insert(_meta._collections.begin(), _meta._collections.end());
-    validateLinks(collections, *vocbase, *this); // remove invalid cids (no such collection or no such link)
-    _meta._collections = std::move(collections);
-  }
 
-  return result;
+  return updateLinks(collections, *vocbase, *this, builder.slice());
 }
 
 /*static*/ std::shared_ptr<LogicalView> IResearchView::make(
     TRI_vocbase_t& vocbase,
     arangodb::velocypack::Slice const& info,
     bool isNew
 ) {
-  auto const id = readViewId(info);
-
-  if (0 == id) {
-    // invalid ID
-    LOG_TOPIC(WARN, IResearchFeature::IRESEARCH)
-      << "got invalid view identifier while constructing IResearch view";
-    return nullptr;
-  }
-
   auto* feature =
     arangodb::iresearch::getFeature<arangodb::DatabasePathFeature>("DatabasePath");
 
   if (!feature) {
     LOG_TOPIC(WARN, IResearchFeature::IRESEARCH)
-      << "failure to find feature 'DatabasePath' while constructing IResearch view '"
-      << id << "' in database '" << vocbase.id() << "'";
+      << "failure to find feature 'DatabasePath' while constructing IResearch View in database '" << vocbase.id() << "'";
 
     return nullptr;
   }
 
-  // get base path from DatabaseServerFeature (similar to MMFilesEngine)
-  // the path is hardcoded to reside under:
-  // <DatabasePath>/<IResearchView::type()>-<view id>
-  // similar to the data path calculation for collections
-  irs::utf8_path dataPath(feature->directory());
-  static std::string subPath("databases");
-
-  dataPath /= subPath;
-  dataPath /= arangodb::iresearch::IResearchView::type().name();
-  dataPath += "-";
-  dataPath += std::to_string(id);
-
-  auto view = std::shared_ptr<IResearchView>(
-    new IResearchView(&vocbase, info, std::move(dataPath), isNew)
-  );
-
-  auto props = info.get("properties");
-
-  if (props.isNone()) {
-    // if no 'properties' then assume defaults
-    props = emptyObjectSlice();
-  }
-
+  PTR_NAMED(IResearchView, view, &vocbase, info, *feature, isNew);
+  auto& impl = reinterpret_cast<IResearchView&>(*view);
+  auto& json = info.isObject() ? info : emptyObjectSlice(); // if no 'info' then assume defaults
+  auto props = json.get("properties");
+  auto& properties = props.isObject() ? props : emptyObjectSlice(); // if no 'info' then assume defaults
   std::string error;
 
-  if (!view->_meta.init(props, error)) {
-    LOG_TOPIC(WARN, iresearch::IResearchFeature::IRESEARCH)
+  if (!impl._meta.init(properties, error)) {
+    LOG_TOPIC(WARN, IResearchFeature::IRESEARCH)
       << "failed to initialize iResearch view from definition, error: " << error;
 
     return nullptr;
   }
 
-  return view;
+  return std::move(view);
```

Comment:
do not use explicit move, it implicitly disable RVO

Reply:
this code was there for a quite long while

Reply:
You guys should compile with clang more often, I get quite a few warnings about places with `return std::move(...)`